lib/ex_plasma/transaction/type/payment_v1.ex

```diff
@@ -1,150 +1,251 @@
-  					65, 226, 241, 55, 0, 110>>
-  			},
-  			output_id: nil,
-  			output_type: 1
-  		}
-  	],
-  	sigs: [],
-  	tx_data: 0,
-  	tx_type: 1
-  }
+  Only validates that the RLP is structurally correct.
+  Does not perform any other kind of validation, use validate/1 for that.
   """
-  @impl Transaction
-  @spec to_map(list()) :: Transaction.t()
-  def to_map(rlp) when is_list(rlp), do: do_to_map(rlp)
-
-  defp do_to_map([sigs, tx_type, inputs, outputs, "", metadata]),
-    do: do_to_map([sigs, tx_type, inputs, outputs, 0, metadata])
-
-  defp do_to_map([sigs, <<tx_type>>, inputs, outputs, tx_data, metadata]),
-    do: do_to_map([sigs, tx_type, inputs, outputs, tx_data, metadata])
-
-  defp do_to_map([sigs, tx_type, inputs, outputs, tx_data, metadata]) do
-    %ExPlasma.Transaction{
-      sigs: sigs,
-      tx_type: tx_type,
-      inputs: Enum.map(inputs, &ExPlasma.Output.decode_id/1),
-      outputs: Enum.map(outputs, &ExPlasma.Output.decode/1),
-      tx_data: tx_data,
-      metadata: metadata
-    }
+  @spec to_map(PaymentV1.t(), list()) :: {:ok, PaymentV1.t()} | {:error, mapping_error()}
+  def to_map(%PaymentV1{}, [<<@tx_type>>, inputs_rlp, outputs_rlp, tx_data_rlp, metadata_rlp]) do
+    with {:ok, inputs} <- decode_inputs(inputs_rlp),
+         {:ok, outputs} <- decode_outputs(outputs_rlp),
+         {:ok, tx_data} <- decode_tx_data(tx_data_rlp),
+         {:ok, metadata} <- decode_metadata(metadata_rlp) do
+      {:ok,
+       %PaymentV1{
+         tx_type: @tx_type,
+         inputs: inputs,
+         outputs: outputs,
+         tx_data: tx_data,
+         metadata: metadata
+       }}
+    end
   end
 
+  def to_map(_, _), do: {:error, :malformed_transaction}
+
+  defp decode_inputs(inputs_rlp) do
+    {:ok, Enum.map(inputs_rlp, &Output.decode_id(&1))}
+  rescue
+    _ -> {:error, :malformed_inputs}
+  end
+
+  defp decode_outputs(outputs_rlp) do
+    {:ok, Enum.map(outputs_rlp, &Output.decode(&1))}
+  rescue
+    _ -> {:error, :malformed_outputs}
+  end
+
+  defp decode_metadata(metadata_rlp) when is_metadata(metadata_rlp), do: {:ok, metadata_rlp}
+  defp decode_metadata(_), do: {:error, :malformed_metadata}
+
+  defp decode_tx_data(@empty_tx_data), do: {:ok, @empty_tx_data}
+
+  defp decode_tx_data(tx_data_rlp) do
+    case RlpDecoder.parse_uint256(tx_data_rlp) do
+      {:ok, tx_data} -> {:ok, tx_data}
+      _ -> {:error, :malformed_tx_data}
+    end
+  end
+
+  @spec get_inputs(PaymentV1.t()) :: list(Output.t())
+  def get_inputs(%PaymentV1{} = transaction), do: transaction.inputs
+
+  @spec get_outputs(PaymentV1.t()) :: list(Output.t())
+  def get_outputs(%PaymentV1{} = transaction), do: transaction.outputs
+
+  @spec get_tx_type(PaymentV1.t()) :: pos_integer()
+  def get_tx_type(%PaymentV1{} = transaction), do: transaction.tx_type
+
   @doc """
   Validates the Transaction.
+  """
+  @spec validate(PaymentV1.t()) :: :ok | {:error, validation_error()}
+  def validate(%PaymentV1{} = transaction) do
```

Comment:
perhaps you should extract the payment_v1 validation stuff?

Reply:
Moved validation to its own module: https://github.com/omgnetwork/ex_plasma/pull/67/commits/811962ded6497d9cb38ff9f6d4b778e571ad6ee6